rust/connlib/snownet/src/allocation.rs

```diff
@@ -1189,6 +1178,22 @@ impl ChannelBindings {
         Some(channel)
     }
 
+    /// Picks the next channel number to allocate.
+    fn next_channel_number(&self) -> u16 {
+        self.inner
+            .keys()
+            .max()
+            .copied()
+            .map(|c| {
+                if c == Self::LAST_CHANNEL {
+                    Self::FIRST_CHANNEL // Wrap around if we have use all channels.
```

Comment:
Do we clear the map at some point? otherwise this will start giving us `FIRST_CHANNEL` after the first time it's filled.

I mean it would still work because we try with increasing channel number but wouldn't that lead to #5603 again?

Reply:
> I mean it would still work because we try with increasing channel number but wouldn't that lead to #5603 again?

Once we cycled through all 4095 channels, it is highly unlikely that the first one is within a clock-drift time of expiry. That would mean we have bound 4095 channels in less than 15 minutes. And if that is the case, we are actually out of channels and there is nothing we can do other than wait.

Reply:
The bug was triggered because we would always try to rebind the last bound channels first and not move to the next one. See the newly added unit test.

Reply:
> Do we clear the map at some point? otherwise this will start giving us `FIRST_CHANNEL` after the first time it's filled.

We clear the map when the allocation fails because that is when it is cleared on the relay.

Reply:
But let's say the map gets filled up to `LAST_CHANNEL`. Now as long as `LAST_CHANNEL` is still used all next binding tries will keep trying to use `FIRST_CHANNEL` first and go sequentially.

The unit tests doesn't trigger that case because they don't fill up the map I believe

Reply:
> But let's say the map gets filled up to `LAST_CHANNEL`. Now as long as `LAST_CHANNEL` is still used all next binding tries will keep trying to use `FIRST_CHANNEL` first and go sequentially.
> 
> The unit tests doesn't trigger that case because they don't fill up the map I believe

Yeah but that is the behaviour we want! If we filled up the map until `LAST_CHANNEL`, `FIRST_CHANNEL` is the one that is most likely expired and can be rebound. In the unlikely event that it isn't, we go to the next one.

We do have a test that checks that we cycle around once the map is full.

I can't see any bugs here, can you? Once the map is full, it is not quite as efficient as it could be. I guess we could add a `handle_timeout` function and proactively clean up the map?

Reply:
wait, it's easier to discuss over a test! pushing one now, feel free to revert!